### DISCUSSION TITLE
Retrieve User Feedback (Thumbs Up/Down) from Open Web UI Pipelines for RLHF

### DISCUSSION THREAD

**Discussion by latent-variable** [open-webui/open-webui#4551] [General]
**Is your feature request related to a problem? Please describe.**
I'm trying to figure out how to retrieve user feedback submitted via the thumbs up/thumbs down interface in the Web UI. Specifically, I need to know how to access this feedback data through the pipelines. Despite searching through the documentation and codebase, I haven't found clear instructions or examples on how to do this.

From what https://github.com/open-webui/pipelines/issues/222 I been told those features are not yet fully implemented 

**Describe the solution you'd like**
I would like to see a feature implemented that allows for the retrieval of user feedback (thumbs up/down) submitted through the Web UI. This feature should enable developers to access and process this feedback data through the pipelines, making it easier to incorporate user feedback into models, especially for tasks like Reinforcement Learning from Human Feedback (RLHF).

**Describe alternatives you've considered**
An alternative could be manually logging this feedback to an external database or file system outside of the pipeline framework, but this approach lacks the seamless integration and ease of use that a built-in feature would provide. Another alternative could involve modifying the existing UI code to expose the feedback, but this would require custom development work and might not be as robust or maintainable.

**Additional context**
The ability to retrieve user feedback directly through the pipelines would be beneficial for continuous improvement of models based on user interactions. This feature could support more effective RLHF workflows, improving model accuracy and user satisfaction. Here’s a screenshot of the current thumbs up/thumbs down interface for reference:
<img width="787" alt="image" src="https://github.com/user-attachments/assets/ab64fdc5-0979-4da9-a999-0d2c7e58eb34">

**Comment by stevenbaert** [comment]
Great input. Is there an update yet? Why would you have thumbs up/down enabled if you cannot gather the output? Has to be there already somewhere (the possibility to gather the info)?

**Comment by aleprj** [comment]
As I can see, it`s stored in the json on the chat column of the chat table. This column is stored as a text column instead of json (at least in Postgresql), so I can`t even query them.
Update: turns out you can.
Here`s the SQL for retrieving (when using Postgresql):

```
select u.email as email, 
	message.value::json->>'model' as model,
	to_timestamp((message.value::json->>'timestamp')::bigint) as datetime,
	message.value::json->'annotation'->>'rating' as rating,
	message.value::json->'annotation'->>'reason' as reason,
	message.value::json->'annotation'->>'comment' as comment,
	t.chat::json->'history'->'messages'->(message.value::json->>'parentId')->>'content' as question,
	message.value::json->>'content' AS answer
	from chat t cross join lateral json_each(t.chat::json#>'{history, messages}') as message
	inner join public.user u on t.user_id = u.id
where message.value::json->'annotation' is not null
order by datetime desc;
```

**Comment by lynxul25** [comment]
Thumbsup for this one. It would allow refinement of system prompts that work best for the users and then increase engagement. It would also help running A/B testing.

**Comment by ChooJunIan** [comment]
I was wondering if the feedback's post request can be sent to a custom API instead of being stored in the chat's database. Does anyone know if this feature has been implemented?

**Comment by alexis-dioxycle** [comment]
It would be great to have that indeed!
